### DISCUSSION TITLE
Building Ledger Live [ledobe] | Beginner walk-through

### DISCUSSION THREAD

**Discussion by nodegraphics** [antichris/ledobe#7] [Q&A] [answered]
This might be a noob question, but im new to docker and trying to run Ledger Desktop on Manjaro [radxa zero] 
I did follow wiki and after "make" comment it went through  and created -desktop and -core folders.
But im wondering after this process where am i gonna find file to run Ledger Live Desktop itselff ? :) 
I cant seem to find it....an dont know where to look.

as a note; after "make" build i did get a error at the end. not sure if it would efffects the overall process...

```
mkdir -p /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/
touch /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/src
mkdir -p /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/
touch /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/patch
mkdir -p /home/codebook/Documents/ledobe-0.3.0/docker/build/core/lib/build
cmake -DBUILD_TESTS=OFF -S /home/codebook/Documents/ledobe-0.3.0/docker/build/core/lib -B /home/codebook/Documents/ledobe-0.3.0/docker/build/core/lib/build
mkdir -p /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/
touch /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/cmake
/bin/sh: line 2: cmake: command not found
make: *** [Makefile:114: /home/codebook/Documents/ledobe-0.3.0/docker/build/.dep/coreLib/cmake] Error 127
[root@zero build]# 
```

Thank you

**Comment by antichris** [comment, marked as answer]
That does not look right at all.

0. First, please, clean up what you just accidentally made in `/home/codebook/Documents/ledobe-0.3.0/docker/build` directory.
    If you used Git do download the Ledobe sources, simply run `git clean -fd`. Otherwise, delete everything and start from a fresh unzipped source.

1. Then, follow the procedure outlined in the [README], and build the Docker image (the easiest is running the bundled `bin/build` script).

2. Once the Docker image has been built successfully, run `make` in a Docker container started from the image. The simplest way to do so is (just like the README has it) by executing
    ```shell
    bin/run make
    ```
    <details><summary>If you wish to make your life just that little bit harder (or know what you're doing)...</summary>

    ... You can execute `bin/run` (or start a container manually, like it says in the README) to get a shell prompt in the container, and then manually run `make` in that environment. Or do whatever else you want to do in the Ledobe container, since it has all the tools necessary for building the components that the Ledger Live Desktop application needs.</details>

When (and if) the build successfully completes, look for the build artifacts (including AppImage) under `desktop/dist` in your build directory (which `bin/run` creates in the Ledobe base directory by default).

You would probably also benefit from reading up the README's of the relevant Ledger Live components (ours has [links] to their GitHub repos), so that you have at least some concept of what it is that should get built and what is involved in that process, in case something goes wrong and you have to debug.

## Note

In general, before you start trying to use any piece of software, it is highly recommended that you familiarize yourself with the README first (and follow the instructions, if there are any). This project is, unsurprisingly, not an exception.

[README]: https://github.com/antichris/ledobe/blob/0.3.0/README.md#process
[links]: https://github.com/antichris/ledobe/blob/0.3.0/README.md#make-targets

**Comment by nodegraphics** [comment]
Thank you  for explanation. iIm gonna give a try for sure.... I dont have much knowledge with docker or even linux environment itself....But im trying to understand / learn along the way.

0- I did delete everything and started from scratch . Downloaded source zip file to "downloads" folder 

1- Extracted ledobe-0.3.0 folder to my "Documents" folder

2- I went inside ledobe-0.3.0 folder and opened my terminal in place and ran `bin/build`  > [build.txt](https://github.com/antichris/ledobe/files/8241764/build.txt)

3- In  same console [after build successfully complete]  i did enter `bin/run make `> [run make.txt](https://github.com/antichris/ledobe/files/8243015/run.make.txt)


-----I did wait couple of hours, then i get this messages....

3a- I did have an unexpected error about network connection [i did check but there wasn't problem with my connection]

```
warning @ledgerhq/ledger-core > node-pre-gyp@0.17.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
[3/5] Fetching packages...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
error An unexpected error occurred: "https://registry.yarnpkg.com/rxjs-compat/-/rxjs-compat-6.6.7.tgz: ESOCKETTIMEDOUT".
info If you think this is a bug, please open a bug report with the information provided in "/build/desktop/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```


3b- After few minutes it finished and gave this error

```
info There appears to be trouble with your network connection. Retrying...
info There appears to be trouble with your network connection. Retrying...
make: *** [Makefile:59: /build/desktop/dist/ledger-live-desktop-2.37.2-linux-arm64.AppImage] Error 1
[codebook@zero ledobe-0.3.0]$ 

```


I did check `desktop/dist `under my ledobe-0.3.0 folder but cant find the `dist `folder....looks like its not there.

Im including screencap of my folder structure;

![Screenshot_2022-03-14_01-41-53](https://user-images.githubusercontent.com/10071439/158135974-f3bc3337-bb8d-4700-8e5f-db1e73a31ed5.png)

![Screenshot_2022-03-14_01-42-53](https://user-images.githubusercontent.com/10071439/158135996-12147ca3-6111-4963-8461-198acced6eac.png)


I hope im making progress . most of the stuff i dont really understand but trying my best.......

Thank you

**Reply by antichris** [reply]
Well, from the message "There appears to be trouble with your network connection", I think it is safe to assume that you had an issue with the network connection there at some point. That [rxjs tgz] downloaded fine for me just now, so I am fairly certain that you would succeed, if you were to retry. And by "retry" I mean, run the same command with `make` again.

P.S.: I would recommend (for your own comfort) that you don't fully redirect the output, but `tee` it instead, e.g.:
```shell
bin/run make | tee run.make.txt
```
As this would allow you to monitor the progress and terminate (`Ctrl+C`) early if you see that things are stuck at some point.

[rxjs tgz]: https://registry.yarnpkg.com/rxjs-compat/-/rxjs-compat-6.6.7.tgz

**Reply by antichris** [reply]
Also, yes, you are making good progress, and things should happen a bit faster at this point, since you've already successfully built the core library (that's the most time consuming part, compiling C++ is like that), now all that is left is building the desktop application itself.

**Reply by nodegraphics** [reply]
Just before your reply , i did ran the command `bin/run make` again after refreshing my connection.....and it did seem to continue where it left.
i got some warnings but it is configuring and packaging i think [ i cant copy the terminal but i got screenshot]

![Screenshot_2022-03-14_02-14-45](https://user-images.githubusercontent.com/10071439/158141412-b4c00465-babe-409c-b1ff-bd493b193a7e.png)

**Reply by nodegraphics** [reply]
ok, my terminal cursor is keep blinking around `waiting` section, its been awhile and im wondering if this is normal ? I can wait but i wanted to double check....
Or should i just cancel and run the command you mentioned above `bin/run make | tee run.make.txt`

**Reply by antichris** [reply]
Does anything interesting happen when you hit `Enter`? If you just get a new line, it is probably still working on something, if you get your shell prompt, it evidently crashed.

**Reply by nodegraphics** [reply]
When i hit `Enter` -waiting....keeps appearing....nothing else....
![Screenshot_2022-03-14_02-50-03](https://user-images.githubusercontent.com/10071439/158147838-d5488963-9d54-4f55-9261-01435d6f2ee3.png)

**Reply by antichris** [reply]
Well, I can't tell for sure, but, since it still does its progress animation, we can only assume it is working on something. Maybe pour yourself a drink, pick up a book or an article to read, or go for a short walk and check back on it later.

I'm also currently running a build to check whether stuff still works as I last left it, and, if I find that something needs fixing, I'll definitely make a new release.

One thing I noticed just moments ago is that they (LedgerHQ) have two different versions of core lib NPM bindings tagged as 6.14.5 ([6.14.5] and [v6.14.5], spot the difference :sweat_smile:), this might be a source of issues. Besides, the apt packages installed in the Docker image also have had some updates, so, in all likelihood, a new Ledobe release is imminent after all.

[6.14.5]: https://github.com/LedgerHQ/lib-ledger-core-node-bindings/blob/6.14.5/package.json#L4
[v6.14.5]: https://github.com/LedgerHQ/lib-ledger-core-node-bindings/blob/v6.14.5/package.json#L4

**Reply by nodegraphics** [reply]
it just passed that waiting phase :)  I see success there so seems like itis all good :)) so far.....

```
[-/23] ⠈ waiting...
success Saved lockfile.
$ node ./scripts/post-install.js
running
no previous checksum saved, will rebuild native deps and save new checksum
creating a new file with checksum
file created
yarn run v1.22.17
$ rimraf node_modules/paralleljs/package.json
$ electron-builder install-app-deps
  • electron-builder  version=22.13.1
  • loaded configuration  file=/build/desktop/electron-builder.yml
  • rebuilding native dependencies  dependencies=@ledgerhq/ledger-core@7.1.0-beta.0, blake-hash@2.0.0, bufferutil@4.0.5, keccak@1.4.0, node-hid@2.1.1, secp256k1@4.0.2, tiny-secp256k1@1.1.6, usb-detection@4.13.0, utf-8-validate@5.0.7, keccak@3.0.0
                                    platform=linux
                                    arch=arm64
  • install prebuilt binary  name=usb-detection version=4.13.0 platform=linux arch=arm64 napi=
  • install prebuilt binary  name=node-hid version=2.1.1 platform=linux arch=arm64 napi= 
  • build native dependency from sources  name=usb-detection
                                          version=4.13.0
                                          platform=linux
                                          arch=arm64
                                          napi=
                                          reason=prebuild-install failed with error (run with env DEBUG=electron-builder to get more information)
                                          error=prebuild-install info begin Prebuild-install version 6.1.4
    prebuild-install WARN install prebuilt binaries enforced with --force!
    prebuild-install WARN install prebuilt binaries may be out of date!
    prebuild-install info looking for local prebuild @ prebuilds/usb-detection-v4.13.0-electron-v89-linux-arm64.tar.gz
    prebuild-install info looking for cached prebuild @ /home/ledobe/.npm/_prebuilds/f81e39-usb-detection-v4.13.0-electron-v89-linux-arm64.tar.gz
    prebuild-install http request GET https://github.com/MadLittleMods/node-usb-detection/releases/download/v4.13.0/usb-detection-v4.13.0-electron-v89-linux-arm64.tar.gz
    prebuild-install http 404 https://github.com/MadLittleMods/node-usb-detection/releases/download/v4.13.0/usb-detection-v4.13.0-electron-v89-linux-arm64.tar.gz
    prebuild-install WARN install No prebuilt binaries found (target=13.5.2 runtime=electron arch=arm64 libc= platform=linux)
    
  • build native dependency from sources  name=node-hid
                                          version=2.1.1
                                          platform=linux
                                          arch=arm64
                                          napi= 
                                          reason=prebuild-install failed with error (run with env DEBUG=electron-builder to get more information)
                                          error=prebuild-install info begin Prebuild-install version 6.1.4
    prebuild-install WARN install prebuilt binaries enforced with --force!
    prebuild-install WARN install prebuilt binaries may be out of date!
    prebuild-install info looking for local prebuild @ prebuilds/node-hid-v2.1.1-napi-v3-linux-arm64.tar.gz
    prebuild-install info looking for cached prebuild @ /home/ledobe/.npm/_prebuilds/7755e1-node-hid-v2.1.1-napi-v3-linux-arm64.tar.gz
    prebuild-install http request GET https://github.com/node-hid/node-hid/releases/download/v2.1.1/node-hid-v2.1.1-napi-v3-linux-arm64.tar.gz
    prebuild-install http 404 https://github.com/node-hid/node-hid/releases/download/v2.1.1/node-hid-v2.1.1-napi-v3-linux-arm64.tar.gz
    prebuild-install WARN install No prebuilt binaries found (target=3 runtime=napi arch=arm64 libc= platform=linux)
    
  • rebuilding native dependency  name=blake-hash version=2.0.0
  • rebuilding native dependency  name=@ledgerhq/ledger-core version=7.1.0-beta.0
  • rebuilding native dependency  name=bufferutil version=4.0.5
  • rebuilding native dependency  name=keccak version=1.4.0
  • rebuilding native dependency  name=node-hid version=2.1.1
  • rebuilding native dependency  name=secp256k1 version=4.0.2
  • rebuilding native dependency  name=tiny-secp256k1 version=1.1.6
  • rebuilding native dependency  name=usb-detection version=4.13.0
  • rebuilding native dependency  name=utf-8-validate version=5.0.7
  • rebuilding native dependency  name=keccak version=3.0.0

```


when it comes to your `6.14.5 vs v6.14.5 ` difference is way over my level of understanding :)
but couple of question that i would like to ask;

1) If the release is imminent, then it means i have to rebuild everything ? Does docker  enviornment has update possibility.
2) if i need to rebuild / compile, i hope the process doesnt change...[please say no]
3) is there any way to distribute this ready build package so that we dont need to go through compiling again ?

**Reply by antichris** [reply]
1. You would have to run `bin/build` to update the Docker image. You'd also have to run `bin/run make` again if the Ledger Live Desktop application and/or component versions would have changed and they would require rebuilding. Fixing this 6.14.5 vs v6.14.5 blunder would do precisely that, so the answer is "yes". I'd also like to update LLD to the latest version (2.39.2, Ledobe currently builds 2.37.2 out of the box)

    Docker images are commonly distributed from an online repository that you can pull the latest (or a specific, of your choosing) version from. Ledobe at this point is not, because of the reasons below (Q/A number 3).

2. At this moment I don't have a reason to make changes to the workflow. So, no, it should stay the same: build the Docker image, start a container from the image and run `make` in that container.

3. Since Ledger Live is extremely security sensitive (it deals with finances, not just private data), I believe the process of originating the LLD application (and its components) should be as transparent as possible. So, even though the Docker image and/or the final LLD ARM64 executable could be distributed ready-built from an online source, unless it is done by the official LedgerHQ, I don't think that's really such a great idea.

    I certainly wouldn't trust my finances to an app I just downloaded from some random guy on the Internet. Now, if I can audit (or ask someone else who can) every step of how my security sensitive app gets built, that's a different story.

**Reply by nodegraphics** [reply]
yes, i understand. it all makes sense. thank you.
looks like all is done and ready to go. >  [run make part2.txt](https://github.com/antichris/ledobe/files/8247745/run.make.part2.txt)

**Comment by nodegraphics** [comment]
**First....I should mention that; after all those process of compile + build, its  great to see Ledger Live interface :)** 

I believe next problem im facing is a known issue.
I did connect my Nano S > Device Detected > Connection failed

![Screenshot_2022-03-14_12-01-18](https://user-images.githubusercontent.com/10071439/158242821-9bc4ce3e-d8c1-4593-9045-0ff40e6e440b.png)

**UPDATE + SOLUTION**

Changing / Installling `udev` rules allows me to connect Nano S + X devices without any problem.

[AUR package |  `ledger-udev 1-7`](https://aur.archlinux.org/packages/ledger-udev)

![Screenshot_2022-03-14_22-56-21](https://user-images.githubusercontent.com/10071439/158316298-22494bd2-d974-4bab-8acc-a4e67a989a6f.png)




@antichris  Thank you for your  patience and for walking me through this process.
